src/map.rs

```diff
@@ -0,0 +1,299 @@
+/// assert_eq!(syscalls.get(Sysno::close), Some(&4));
+/// assert_eq!(syscalls.insert(Sysno::close, 11), Some(4));
+/// assert_eq!(syscalls.count(), 1);
+/// assert_eq!(syscalls.remove(Sysno::close), Some(11));
+/// assert!(syscalls.is_empty());
+/// ```
+#[derive(Clone, Eq, PartialEq)]
+pub struct SysnoMap<T> {
+    is_set: SysnoSet,
+    data: [T; Sysno::table_size()],
+}
+
+impl<T: Default> SysnoMap<T> {
+    /// Initialize an empty syscall map, must have hardcoded size `Sysno::table_size()`.
+    pub fn new(data: [T; Sysno::table_size()]) -> Self {
+        Self {
+            is_set: SysnoSet::empty(),
+            data,
+        }
+    }
+
+    /// Initialize an syscall map for the given syscalls.
+    pub fn new_with_set(
+        data: [T; Sysno::table_size()],
+        sysno_set: SysnoSet,
+    ) -> Self {
+        Self {
+            is_set: sysno_set,
+            data,
+        }
+    }
+
+    /// Returns true if the map contains the given syscall.
+    pub const fn contains_key(&self, sysno: Sysno) -> bool {
+        self.is_set.contains(sysno)
+    }
+
+    /// Clears the map, removing all syscalls.
+    pub fn clear(&mut self) {
+        self.is_set.clear();
+        for elem in self.data.iter_mut() {
+            *elem = T::default();
+        }
+    }
+
+    /// Returns true if the map is empty. This is an O(n) operation as
+    /// it must iterate over the entire bitset.
+    pub fn is_empty(&self) -> bool {
+        self.is_set.is_empty()
+    }
+
+    /// Returns the number of syscalls in the map. This is an O(n) operation as
+    /// it must count the number of bits in the bitset.
+    pub fn count(&self) -> usize {
+        self.is_set.count()
+    }
+
+    /// Inserts the given syscall into the map. Returns true if the syscall was not already in the map.
+    pub fn insert(&mut self, sysno: Sysno, value: T) -> Option<T> {
+        let old = mem::replace(&mut self.data[data_idx(sysno)], value);
+        if self.is_set.insert(sysno) {
+            None
+        } else {
+            Some(old)
+        }
+    }
+
+    /// Removes the given syscall from the map. Returns true if the syscall was in the map.
+    pub fn remove(&mut self, sysno: Sysno) -> Option<T> {
+        if self.is_set.remove(sysno) {
+            Some(mem::take(&mut self.data[data_idx(sysno)]))
+        } else {
+            None
+        }
+    }
+
+    #[inline]
+    pub fn get(&self, sysno: Sysno) -> Option<&T> {
+        if self.is_set.contains(sysno) {
+            Some(&self.data[data_idx(sysno)])
+        } else {
+            None
+        }
+    }
+
+    #[inline]
+    pub fn get_mut(&mut self, sysno: Sysno) -> Option<&mut T> {
+        if self.is_set.contains(sysno) {
+            Some(&mut self.data[data_idx(sysno)])
+        } else {
+            None
+        }
+    }
+
+    /// Returns an iterator that iterates over the syscalls contained in the map.
+    pub fn iter(&self) -> impl Iterator<Item = (Sysno, &T)> {
```

Comment:
Consider the case where the user may want to write some code like this:
```rust
struct MyIterAdaptor<T> {
    iter: SysnoMapIter<T>,
}
```

If `SysnoMapIter` didn't exist, then there would be no way to do this because it isn't possible to name the type of `impl Iterator`.